docs/proposed/psa-driver-interface.md

```diff
@@ -321,6 +321,81 @@ TODO: key input and output for opaque drivers; deterministic key generation for
 
 TODO
 
+#### PAKE operation driver dispatch logic
+
+PSA PAKE operation structure for driver dispatch:
+
+```
+struct psa_pake_operation_s
+{
+    /** Unique ID indicating which driver got assigned to do the
+     * operation. Since driver contexts are driver-specific, swapping
+     * drivers halfway through the operation is not supported.
+     * ID values are auto-generated in psa_crypto_driver_wrappers.h
+     * ID value zero means the context is not valid or not assigned to
+     * any driver (i.e. none of the driver contexts are active). */
+    unsigned int MBEDTLS_PRIVATE(id);
+    /* Algorithm used for PAKE operation */
+    psa_algorithm_t MBEDTLS_PRIVATE(alg);
+    /* Based on stage (collecting inputs/computation) we select active structure of data union.
+     * While switching stage (when driver setup is called) collected inputs
+       are copied to the corresponding operation context. */
+    uint8_t MBEDTLS_PRIVATE(stage);
+    /* Holds the computation stage of the PAKE algorithms. */
+    psa_pake_computation_stage_t MBEDTLS_PRIVATE(computation_stage);
+    union {
+        unsigned dummy;
+        psa_crypto_driver_pake_inputs_t MBEDTLS_PRIVATE(inputs);
+        psa_driver_pake_context_t MBEDTLS_PRIVATE(ctx);
+    } MBEDTLS_PRIVATE(data);
+};
+```
+
+PAKE operation is divided into two stages: `collecting inputs` and `computation`. `stage` field defines the current stage and selects the active structure of the `data` union.
+The core decides whether to dispatch a PAKE operation to a driver based on the location of the provided password while calling `pake_setup` driver entry point.
+The core is responsible for holding information about the current stage of computation(`computation_stage`) and provides this information to the driver.
+
+1. Collecting inputs stage
+
+The core conveys the initial inputs for a PAKE operation via an opaque data structure of type `psa_crypto_driver_pake_inputs_t`.
+After calling `psa_pake_setup` the operation object is initialized and is ready to collect inputs. Driver entry point for `pake_setup` is not called at this point. It will be called later when all inputs are collected. Setter functions: `psa_pake_set_password_key`, `psa_pake_set_role`, `psa_pake_set_user`, `psa_pake_set_peer` do not have driver entry points. These functions just fill `inputs` structure.
+
+2. Computation stage
+
+First call of `psa_pake_output()` or `psa_pake_input()` switches the stage to `computation` (assuming that all inputs are collected) and calls `pake_setup` driver entry point. Driver function is responsible for coping inputs from given `inputs` structure to the driver context. Note that, after calling `pake_setup` the driver entry point, core will free memory allocated for the password. The driver is responsible for making its own copy.
```

Comment:
The driver is not required to store the inputs in the driver context, but it is the likely thing that will happen.

Reply:
I don't think so. Inputs and driver context shares the same memory. `pake_setup` driver entry point is responsible to store what the driver needs from inputs. Then when the driver starts using its context the inputs are destroyed.
`pake_setup` should store what it needs from inputs locally, clear the context (inputs are lost) and fill the context with required data.
`password` is a special case, because it holds address in memory when the password is kept (allocated by core). In the following [discussion](https://github.com/Mbed-TLS/mbedtls/pull/6567#discussion_r1052337538) we decided to free memory for password in core after return from `pake_setup` driver call.

Reply:
> Inputs and driver context shares the same memory.

Not sure what you mean here. Do you mean that they are fields of an union?

Reply:
I have checked in `crypto_extra.h` and this is indeed the case:
```C
union {
        unsigned dummy;
        psa_crypto_driver_pake_inputs_t MBEDTLS_PRIVATE(inputs);
        psa_driver_pake_context_t MBEDTLS_PRIVATE(ctx);
    } MBEDTLS_PRIVATE(data);
```
but then in mbedtls_psa_pake_setup we are going to read and write in the union memory space at the same time or am I missing something?

Reply:
I think it is overly complicated and tricky to do it right when implementing a pake_setup() entry point. Another point is that this saves typically 20 bytes in the core operation context but increases by more than that the code size to handle this complexity. I definitely prefer the solution where the inputs parameters are kept for the lifetime of the operation in the core context where drivers can access them. That way also, the driver does not have to do a copy of the password, that's one less allocation and thus one less potential problem.

Reply:
This is what I've tried to promote here. In #5451, I have not found documentation about the lifetime of the `inputs` parameter though. Here I suggest for it to have the lifetime of the core operation. This enables the driver to just save the pointer (not mandatory though) in `pake_setup` and to access the initial inputs (through getter functions as defined in #5451) as long as the core as not called `pake_abort`. Without that the driver has to copy the initial inputs it needs in its context especially the password.

Reply:
> Another point is that this saves typically 20 bytes in the core operation context but increases by more than that the code size to handle this complexity.

I agree. If we chose to separate inputs and driver context (remove the union), then the driver can only store the pointer to inputs and use them when needed. Moving inputs to driver context is not needed. `stage` can be removed from `psa_pake_operation_s` and code will be less complicated, but it will be inconsistent with solution proposed for key derivation.

Reply:
> This enables the driver to just save the pointer (not mandatory though) in pake_setup and to access the initial inputs (through getter functions as defined in https://github.com/Mbed-TLS/mbedtls/pull/5451) as long as the core has not called pake_abort

It is not clear to me why do we need getter functions for inputs if the structure members can be accessed directly. Is this not overkill? Maybe I'm missing something, but it seems getter function for `role` could look like this:
```
psa_pake_role_t pas_pake_get_role( const psa_crypto_driver_pake_inputs_t *inputs )
{
    return( inputs->role );
}
```
Wouldn't be easier to access it directly: `operation->inputs->role` ?

Reply:
> the lifetime of the inputs parameter

When a function receives a pointer as argument, the default assumption is that the pointer remains valid until the function returns, and is no longer valid after that. This is the case for `inputs`. In key derivation, from the driver's point of view the `inputs` structure is only valid during the execution of the `"key_derivation_setup"` entry point.

Between calls to API functions, there is never any guarantee that a pointer remains valid. A concrete scenario where pointers to the “same” data structure can change is if the library is called from a high-level language and the memory manager (typically a garbage collector) can move memory blocks. In this scenario, pointers are only locked in place during the execution of a function that has access to them.

Reply:
Thanks for good explanation. In this case I think we should keep the union of `inputs` and `ctx`. Proposed solution with copy of inputs in the setup driver wrapper solves the issue with complicated implementation of the setup driver entry point (preserving inputs on stack before filling the context).

Reply:
> > the lifetime of the inputs parameter
> 
> When a function receives a pointer as argument, the default assumption is that the pointer remains valid until the function returns, and is no longer valid after that. This is the case for `inputs`. In key derivation, from the driver's point of view the `inputs` structure is only valid during the execution of the `"key_derivation_setup"` entry point.
> 
> Between calls to API functions, there is never any guarantee that a pointer remains valid. A concrete scenario where pointers to the “same” data structure can change is if the library is called from a high-level language and the memory manager (typically a garbage collector) can move memory blocks. In this scenario, pointers are only locked in place during the execution of a function that has access to them.

I agree with that. I previously forgot about this pointer lifetime constraint.